### PR TITLE
[AIRFLOW-6058] Running tests with pytest

### PR DESCRIPTION
**Pull request by turbaszek**
Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-6058

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:

This PR switches nosetest to pytest. I tried to keep required changes to minimum.
What I have already learned is that:
- many tests depend on database
- order of tests is important
- there are tests that create side effects in logging setup, local settings, config.

I am open to any suggestions and help.

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

### CODE REVIEW

File: tests/conftest.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+
+@pytest.fixture(autouse=True)
+def reset_environment():
+    """
+    Resets env variables.
+    """
+    init_env = os.environ
+    yield
+    changed_env = os.environ
```
Comment (mik-laj) [inline tests/conftest.py:33]: Shouldn't you do deepcopy here?

File: setup.py [hunk 2]
```diff
@@ -296,6 +297,9 @@ def write_version(filename: str = os.path.join(*["airflow", "git_version"])):
     'pylint~=2.3.1',  # to be upgraded after fixing https://github.com/PyCQA/pylint/issues/3123
                       # We should also disable checking docstring at the module level
     'pysftp',
+    'pytest',
+    'pytest-cov==2.8.1',
+    # 'pytest-randomly'
```
Comment (feluelle) [inline setup.py]: So we can get rid of nose :)
```diff
- 'nose',
- 'nose-ignore-docstring==0.2',
- 'nose-timer',
```

File: tests/conftest.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+
+@pytest.fixture()
+def reset_db():
+    """
+    Resets Airflow db.
+    """
+    db.resetdb()
+    yield
+    pass
```
Comment (feluelle) [inline tests/conftest.py]:
```suggestion
```
?

**Comment by feluelle** [comment]
Oh just forgot it is not ready for review, sry 😬

**Comment by feluelle** [comment]
I can't wait to finally have pytest 😁

**Comment by akki** [comment]
Is there any place (JIRA ticket or discussion) where I can know why we want to move towards `pytest`?

**Comment by turbaszek** [comment]
@akki here is mail with AIP-27:
https://lists.apache.org/thread.html/484add15d555c717bab82553ba7e43a4e04f0d403ccf954406b382e7@%3Cdev.airflow.apache.org%3E

and some further discussion:
https://lists.apache.org/thread.html/ad6d7fce82b0a749e21066b6e5c3b80a5d59eca9166be86ceeaea05d@%3Cdev.airflow.apache.org%3E

File: airflow/utils/log/colored_log.py [hunk 4]
```diff
@@ -89,6 +89,9 @@ def _color_record_traceback(self, record: LogRecord) -> LogRecord:
...
     def format(self, record: LogRecord) -> str:
-        record = self._color_record_args(record)
-        record = self._color_record_traceback(record)
-        return super().format(record)
+        try:
+            record = self._color_record_args(record)
+            record = self._color_record_traceback(record)
+            return super().format(record)
+        except ValueError:  # I/O operation on closed file
+            return Formatter().format(record)
```
Comment (feluelle) [inline airflow/utils/log/colored_log.py:98]: Should we import the `Formatter` here? So it will only import if really required.

File: airflow/www/views.py [hunk 5]
```diff
@@ -1024,12 +1024,12 @@ def blocked(self, session=None):
             )
             if 'all_dags' not in filter_dag_ids:
                 dags = dags.filter(DR.dag_id.in_(filter_dag_ids))
-            dags = dags.all()
+            else:
```
Comment (feluelle) [inline airflow/www/views.py]: I am not sure if this is correct. [all()](https://docs.sqlalchemy.org/en/13/orm/query.html#sqlalchemy.orm.query.Query.all) collects the dags - [filter()](https://docs.sqlalchemy.org/en/13/orm/query.html#sqlalchemy.orm.query.Query.filter) does only filter the dags but not collect them.

File: setup.py [hunk 6]
```diff
@@ -299,6 +300,9 @@ def write_version(filename: str = os.path.join(*["airflow", "git_version"])):
...
```
Comment (feluelle) [inline setup.py]: `pytest-randomly` is this useful?

File: setup.py [hunk 7]
```diff
@@ -299,6 +300,9 @@ def write_version(filename: str = os.path.join(*["airflow", "git_version"])):
     'pylint~=2.3.1',  # to be upgraded after fixing https://github.com/PyCQA/pylint/issues/3123
                       # We should also disable checking docstring at the module level
     'pysftp',
+    'pytest',
```
Comment (feluelle) [inline setup.py:386]: I think we should also increase the `DEPENDENCIES_EPOCH_NUMBER` in the Dockerfile, right @potiuk ?

File: airflow/www/views.py [hunk 8]
```diff
@@ -1024,12 +1024,12 @@ def blocked(self, session=None):
             )
             if 'all_dags' not in filter_dag_ids:
                 dags = dags.filter(DR.dag_id.in_(filter_dag_ids))
-            dags = dags.all()
```
Comment (mik-laj) [inline airflow/www/views.py]: Why did you change it? This should work because all is a method of the query class that retrieves all items that are filtered in the current query. This work differently from "one", which only retrieves first. https://docs.sqlalchemy.org/en/13/orm/query.html#sqlalchemy.orm.query.Query.all

File: airflow/www/views.py (hunk 5 above)
Comment (turbaszek) [reply airflow/www/views.py]: Yeah, that's true. Thanks!

File: tests/conftest.py [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+
+from airflow.utils import db
+
+
+@pytest.fixture(autouse=True)
+def reset_environment():
+    """
+    Resets env variables.
+    """
+    init_env = deepcopy(os.environ)
```
Comment (mik-laj) [inline tests/conftest.py]:
```suggestion
    init_env =os.environ.copy()
```
In python standard library, we have a dedicated method for this.

File: airflow/www/views.py [hunk 10]
```diff
@@ -1029,7 +1029,6 @@ def blocked(self, session=None):
             for dag_id, active_dag_runs in dags:
                 max_active_runs = 0
                 dag = dagbag.get_dag(dag_id)
-                max_active_runs = dagbag.dags[dag_id].max_active_runs
```
Comment (potiuk) [inline airflow/www/views.py:1051]: Why removing it here? I guess the next 2 lines make it works as expected?

File: setup.py [hunk 11]
```diff
@@ -285,20 +285,20 @@ def write_version(filename: str = os.path.join(*["airflow", "git_version"])):
     'dumb-init>=1.2.2',
     'flake8>=3.6.0',
     'flake8-colors',
+    'flaky==3.6.1',
     'freezegun',
     'ipdb',
     'jira',
     'mongomock',
     'moto==1.3.5',
-    'nose',
```
Comment (potiuk) [inline setup.py:379]: NICE! Waited for it for a long time!

File: setup.py (hunk 7 above)
Comment (potiuk) [reply setup.py:386]: Correct @feluelle -> @nuclearpinguin, otherwise nose will still be there in the image.

File: tests/conftest.py (hunk 9 above)
Comment (potiuk) [reply tests/conftest.py]: How about using monkeypatch ? https://docs.pytest.org/en/latest/monkeypatch.html#monkeypatching-environment-variables - I guess you wanted to be 100% sure that we will restore all env variables instead of changing all the places where os.environ gets modified?

I checked that there seem to be a handful of places where environ is modified. 
I know it is safer to save the env, but on the other hand it might be a lot slower:

Screenshot 2019-11-25 at 22 39 53

WDYT?

File: tests/pytest.ini [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+[pytest]
+addopts =
+    -rasl
+    --cov=airflow/
+    --cov-report html:airflow/www/static/coverage/
+    --ignore=tests/dags/test_dag_serialization.py
```
Comment (potiuk) [inline tests/pytest.ini]: Why do we ignore dag serialization?

File: tests/task/task_runner/test_standard_task_runner.py [hunk 13]
```diff
@@ -121,11 +121,12 @@ def test_on_kill(self):
...
 
-        # give the task some time to startup
-        time.sleep(3)
-
+        # Give the task some time to startup
+        time.sleep(10)
         runner.terminate()
 
+        # Wait some time for the result
+        time.sleep(40)
```
Comment (potiuk) [inline tests/task/task_runner/test_standard_task_runner.py]: Hmm... Are we sure we cannot do some loop waiting for result - even in an active loop?.

File: tests/test_logging_config.py [hunk 14]
```diff
@@ -157,8 +157,13 @@ def setUp(self):
     def tearDown(self):
         # Remove any new modules imported during the test run. This lets us
         # import the same source files for more than one test.
+        from airflow.logging_config import configure_logging
```
Comment (potiuk) [inline tests/test_logging_config.py:160]: Nice!

File: tests/www/test_views.py [hunk 15]
```diff
@@ -1129,7 +1129,8 @@ class TestDagACLView(TestBase):
     """
     Test Airflow DAG acl
     """
-    default_date = timezone.datetime(2018, 6, 1)
+    next_year = dt.now().year + 1
```
Comment (potiuk) [inline tests/www/test_views.py:1175]: :D -> 2018 !

**Review by potiuk — APPROVED** [review]
Some small comments.

File: tests/pytest.ini (hunk 12 above)
Comment (turbaszek) [reply tests/pytest.ini]: Because the test didn't work. But since DAG serialisation the test does not exist so I will remove it.

File: tests/task/task_runner/test_standard_task_runner.py (hunk 13 above)
Comment (turbaszek) [reply tests/task/task_runner/test_standard_task_runner.py]: You mean a loop that will wait for file to be present? Sounds reasonable.

File: airflow/www/views.py (hunk 10 above)
Comment (turbaszek) [reply airflow/www/views.py:1051]: Yes, sometimes the dag is `None` (as suggested also by the if condition) and in that cases the `dag_id` key may not be present. In general I think  `get_dag` is the right way to get a DAG from dagbag.

File: tests/conftest.py (hunk 9 above)
Comment (turbaszek) [reply tests/conftest.py]: I agree. Instead of monkeypatch we can mark those test with `usefixture` and use the fixture in question, WDYT?

File: setup.py [hunk 16]
```diff
@@ -285,20 +285,20 @@ def write_version(filename: str = os.path.join(*["airflow", "git_version"])):
...
     'paramiko',
     'pre-commit',
     'pylint~=2.3.1',  # to be upgraded after fixing https://github.com/PyCQA/pylint/issues/3123
                       # We should also disable checking docstring at the module level
     'pysftp',
+    'pytest',
+    'pytest-cov==2.8.1',
     'pywinrm',
     'qds-sdk>=1.9.6',
     'rednose',
```
Comment (ashb) [inline setup.py]: Can get rid of this too

File: tests/__init__.py [hunk 17]
```diff
@@ -20,4 +20,4 @@
 # flake8: noqa
 
 from .api import *  # type: ignore
-from .core import *  # type: ignore
+from .test_core import *  # type: ignore
```
Comment (ashb) [inline tests/__init__.py]: Do we still need this line? It was I think only a kludge as our files weren't named as expected, but it now is

File: tests/pytest.ini [hunk 18]
```diff
@@ -0,0 +1,37 @@
...
+    tests/dags_with_system_exit
+    tests/test_utils
+    tests/dags_corrupted
+faulthandler_timeout=180
+log_print = True
+log_level = INFO
+filterwarnings =
+    ignore::DeprecationWarning
+    ignore::PendingDeprecationWarning
+    ignore::RuntimeWarning
```
Comment (ashb) [inline tests/pytest.ini]: I'm not sure about these, but especially this one.

What was your thinking for this?

Comment (turbaszek) [reply tests/pytest.ini]: The general idea was to declutter logs by removing deprecation warnings. I have to check why I added the `RuntimeWarning` but I can agree that this one is quite disputable.

Comment (turbaszek) [reply tests/pytest.ini]: I see no `RuntimeWarning` now. What do you think about filtering out other warnings? @potiuk @mik-laj @ashb @feluelle

File: tests/__init__.py (hunk 17 above)
Comment (turbaszek) [reply tests/__init__.py]: It seems that everything can be removed from this `__init__` ;)

File: tests/conftest.py (hunk 9 above)
Comment (turbaszek) [reply tests/conftest.py]: I checked all those occurrences and fixed `test_hive_hook` in other places it's used in helper or handled properly.

Comment (potiuk) [reply tests/conftest.py]: @nuclearpinguin ^^ monkeypatch seems nicer and pytest-dedicated way of temporary modifying the environment (and other things). It is - far less code and it restores environment automatically. So if we decide to optimise the speed and get-rid of auto-fixture, I think monkeypatch is better solution.

File: airflow/www/views.py (hunk 10 above)
Comment (potiuk) [reply airflow/www/views.py:1051]: 👍

File: tests/pytest.ini (hunk 18 above)
Comment (potiuk) [reply tests/pytest.ini]: I am all ok for filtering Deprecation warnings but only when we run regular PRs in CI. If we run tests locally, we should still see the warnings (that encourages to fix the warnings). I think there is very little value in warnings displayed in CI for regular PRs. Usually those will be the "other" warnings that you are not really interested in. 

Maybe, what could be done instead - we have the CRON build run daily where we already have a more comprehensive tests (for example we build the whole Docker image from the scratch and we always run Kubernetes tests no matter if any of kubernetes files changed). Maybe we could change it that Deprecation warnings are still displayed when you run tests locally and when you run CRON CI job, but not when you run regular PR.

File: setup.py [hunk 19]
```diff
@@ -301,7 +301,6 @@ def write_version(filename: str = os.path.join(*["airflow", "git_version"])):
     'pytest-cov==2.8.1',
     'pywinrm',
     'qds-sdk>=1.9.6',
-    'rednose',
```
Comment (potiuk) [inline setup.py:390]: Right :).

File: tests/pytest.ini [hunk 20]
```diff
@@ -20,7 +20,6 @@ addopts =
     -rasl
     --cov=airflow/
     --cov-report html:airflow/www/static/coverage/
-    --ignore=tests/dags/test_dag_serialization.py
```
Comment (potiuk) [inline tests/pytest.ini]: Good :)

File: tests/task/task_runner/test_standard_task_runner.py [hunk 21]
```diff
@@ -126,7 +126,11 @@ def test_on_kill(self):
         runner.terminate()
 
         # Wait some time for the result
-        time.sleep(40)
+        for _ in range(5):
```
Comment (potiuk) [inline tests/task/task_runner/test_standard_task_runner.py]: Much better !

File: tests/pytest.ini (hunk 18 above)
Comment (feluelle) [reply tests/pytest.ini]: I think we should differ between `DeprecationWarning`s in our Airflow code base and external modules. IMO we should always try to fix deprecation warnings in our code base. And we can ignore the external ones.

Comment (feluelle) [reply tests/pytest.ini]: > And we can ignore the external ones.

Because sometimes those `DeprecationWarnings` of external modules cannot be fixed by us.

Comment (turbaszek) [reply tests/pytest.ini]: I agree with both of you @feluelle and @potiuk. Removing depreciation warning adds around 800 lines to CI log. So I would do as @potiuk suggested: filter warning only when run on CI.

File: scripts/ci/in_container/entrypoint_ci.sh [hunk 22]
```diff
@@ -219,45 +219,31 @@ if [[ "${RUN_TESTS}" == "false" ]]; then
...
-    ARGS=("--with-coverage"
-          "--cover-erase"
-          "--cover-html"
-          "--cover-package=airflow"
-          "--cover-html-dir=airflow/www/static/coverage"
-          "--with-ignore-docstrings"
-          "--rednose"
-          "--with-xunit"
-          "--xunit-file=${XUNIT_FILE}"
-          "--with-timer"
```
Comment (ashb) [inline scripts/ci/in_container/entrypoint_ci.sh:232]: The pytest "equivalent" of this is `--durations=100` (to show timing for slowest 100 tests). Shall we add it back?

File: tests/task/task_runner/test_standard_task_runner.py [hunk 23]
```diff
@@ -121,11 +121,16 @@ def test_on_kill(self):
         runner = StandardTaskRunner(job1)
         runner.start()
 
-        # give the task some time to startup
-        time.sleep(3)
-
+        # Give the task some time to startup
+        time.sleep(10)
```
Comment (ashb) [inline tests/task/task_runner/test_standard_task_runner.py:125]: Why so much more time than before?

**Comment by ashb** [comment]
@nuclearpinguin Looking at the test run time something is going on here -- the tests are now running for ~35-40mins each instead of ~26-28mins. Could you see what's going on?

**Comment by ashb** [comment]
It might be worth adding `--verbose`/`-v` in to the travis args too so we get a bit more info as it goes? Not sure either way on this

**Comment by ashb** [comment]
```
tests/executors/test_executor.py:27

  /opt/airflow/tests/executors/test_executor.py:27: PytestCollectionWarning: cannot collect test class 'TestExecutor' because it has a __init__ constructor (from: executors/test_executor.py)

    class TestExecutor(BaseExecutor):
```

That class is an executor for testing, not tests of an executor, so we should probably rename that. Perhaps put it under tests.test_utils?

Similar for TestSecurityManager class. (Probably renaming is the best answer there)

**Comment by ashb** [comment]
```
Coverage.py warning: Couldn't read data from '/opt/airflow/.coverage.71d1c30656a3.10847.269416': JSONDecodeError: Expecting ',' delimiter: line 1 column 759 (char 758)

Coverage.py warning: Couldn't read data from '/opt/airflow/.coverage.71d1c30656a3.11170.539483': JSONDecodeError: Expecting value: line 1 column 23418 (char 23417)
```

This maybe isn't a problem as coverage report still seems to be uploaded fine.

File: tests/task/task_runner/test_standard_task_runner.py (hunk 23 above)
Comment (turbaszek) [reply tests/task/task_runner/test_standard_task_runner.py:125]: This test is flaky, but now with the loop below I think we can go back to 3s.
Edit: It seems that 10s is better as the test does not fail:
https://travis-ci.org/apache/airflow/jobs/617721621

**Comment by turbaszek** [comment]
> @nuclearpinguin Looking at the test run time something is going on here -- the tests are now running for ~35-40mins each instead of ~26-28mins. Could you see what's going on?

Yeah, since the beginning of work on pytest I was wondering about that. I will add the -v flag and timing but still, if you will observe how tests are run you will see that some are slow. Progres from ~30% to ~85% takes like 4 minutes...

**Comment by turbaszek** [comment]
There are 25 tests that last longer than 10s, the longest test is:
```
48.25s call     cli/commands/test_task_command.py::TestCliTaskBackfill::test_run_ignores_all_dependencies
```
whole test suite according to pytest was run under 30min:
```
==== 1 failed, 3985 passed, 117 skipped, 16 warnings in 1599.60s (0:26:39) =====
```
and the whole [build](https://travis-ci.org/PolideaInternal/airflow/jobs/617721613) from which I took those statistics took `40 min 4 sec`.

Comparing with present [master](https://travis-ci.org/apache/airflow/jobs/617025719) the difference looks like this:
- 1599.60s > 1360.767s (5 minutes)
- 3986  > 3981 run tests (5 tests)

File: scripts/ci/in_container/entrypoint_ci.sh (hunk 22 above)
Comment (turbaszek) [reply scripts/ci/in_container/entrypoint_ci.sh:232]: Done

File: tests/pytest.ini (hunk 18 above)
Comment (feluelle) [reply tests/pytest.ini]: I am fine with that. :) 👍

File: setup.py (hunk 7 above)
Comment (feluelle) [reply setup.py:386]: Actually.. Can we not just compare the dependencies before and after to check if we need to rebuild?

File: tests/task/task_runner/test_standard_task_runner.py [hunk 24]
```diff
@@ -121,11 +121,16 @@ def test_on_kill(self):
...
         runner.terminate()
 
+        # Wait some time for the result
+        for _ in range(5):
+            if os.path.exists(path):
+                break
+            time.sleep(8)
```
Comment (houqp) [inline tests/task/task_runner/test_standard_task_runner.py]: 8 seconds seems like a lot, i would recommend increase the range count and decrease the sleep interval here.

File: tests/task/task_runner/test_standard_task_runner.py (hunk 23 above)
Comment (potiuk) [reply tests/task/task_runner/test_standard_task_runner.py]: The problem is with wrong dependencies epoch number change. APT-deps are re-installed from the scratch in this PR which makes it take longer. I will explain it in comment to the PR.

File: Dockerfile [hunk 25]
```diff
@@ -35,7 +35,7 @@ ENV DEBIAN_FRONTEND=noninteractive LANGUAGE=C.UTF-8 LANG=C.UTF-8 LC_ALL=C.UTF-8
     LC_CTYPE=C.UTF-8 LC_MESSAGES=C.UTF-8
 
 # By increasing this number we can do force build of all dependencies
-ARG DEPENDENCIES_EPOCH_NUMBER="2"
+ARG DEPENDENCIES_EPOCH_NUMBER="3"
```
Comment (potiuk) [inline Dockerfile]: Those are wrong dependencies changed. PIP_DEPENDENCIES_EPOCH_NUMBER should be changed (which is way lower in the Docker file. That's causing much longer builds in Travis (until merged).

**Review by potiuk — CHANGES_REQUESTED** [review]
Performance is hampered by changing different EPOCH number. 
We also need to update the script that prepares list of test cases for `run-test` script:
/Users/potiuk/code/airflow/scripts/ci/in_container/run_extract_tests.sh

**Comment by potiuk** [comment]
> @nuclearpinguin Looking at the test run time something is going on here -- the tests are now running for ~35-40mins each instead of ~26-28mins. Could you see what's going on?

@ashb -> that's about wrong EPOCH change in Dockerfile. Should be PIP_EPOCH

**Comment by ashb** [comment]
> > @nuclearpinguin Looking at the test run time something is going on here -- the tests are now running for ~35-40mins each instead of ~26-28mins. Could you see what's going on?
> 
> @ashb -> that's about wrong EPOCH change in Dockerfile. Should be PIP_EPOCH

Ahhhhh that would explain it.

**Comment by turbaszek** [comment]
> Performance is hampered by changing different EPOCH number.
> We also need to update the script that prepares list of test cases for `run-test` script:
> /Users/potiuk/code/airflow/scripts/ci/in_container/run_extract_tests.sh

What is purpose of this script @potiuk ? If this is only for autocomplete then I think we have to think it over as pytest support [the powerful -k flag ](http://doc.pytest.org/en/latest/usage.html#specifying-tests-selecting-tests). Plus pytest accept path to tests in normal /-separated way so one can use normal autocomplete .

File: tests/task/task_runner/test_standard_task_runner.py (hunk 24 above)
Comment (turbaszek) [reply tests/task/task_runner/test_standard_task_runner.py]: Agree, fixed

File: Dockerfile (hunk 25 above)
Comment (turbaszek) [reply Dockerfile]: Too much magic 🦄 Fixed!

**Comment by potiuk** [comment]
@nuclearpinguin -> as discussed. I am more than fine with relying on  pytest's feature now allowing to specify path of the file to test + some explanation in BREEZE as well as in the "run-tests" hints about using -k flag and ::test_method. This is so much more straightforward now with pytest

**Comment by turbaszek** [comment]
> @nuclearpinguin -> as discussed. I am more than fine with relying on pytest's feature now allowing to specify path of the file to test + some explanation in BREEZE as well as in the "run-tests" hints about using -k flag and ::test_method. This is so much more straightforward now with pytest

The only custom and useful part of `run-tests` was the `--with-db-init` flag. So I added it as a custom pytest plugin with additional session fixture, this also allowed to simplify some scripts. As a result running tests in Breeze goes down to using pure pytest 🚀 This includes getting help as `pytest --help` includes:
```
airflow:
  --with-db-init        Forces database initialization before tests
```

File: BREEZE.rst [hunk 26]
```diff
@@ -506,40 +506,48 @@ capturing the stdout of your process.
...
+in official documentation but here are few basic examples:
 
 .. code-block:: bash
 
-   run-tests tests.core:TestCore.test_check_operators -- -s --logging-level=DEBUG
+    pytest -k "TestCore and not check"
 
-The tests run ``airflow db reset`` and ``airflow db init`` the first time you
-launch them in a running container, so you can count on the database being initialized.
+The example above will run ``TestCore.test_bash_operator_kill``  but not ``TestMyClass.test_check_operators``.
```
Comment (potiuk) [inline BREEZE.rst]:
```suggestion
The example above will run ``TestCore.test_bash_operator_kill``  but not ``TestCore.test_check_operators``.
```

File: scripts/ci/in_container/entrypoint_ci.sh [hunk 27]
```diff
@@ -33,6 +33,8 @@ in_container_basic_sanity_check
 
 in_container_script_start
 
+TRAVIS=
```
Comment (potiuk) [inline scripts/ci/in_container/entrypoint_ci.sh]:
```suggestion
TRAVIS=${TRAVIS:=}
```

File: tests/conftest.py [hunk 28]
```diff
@@ -0,0 +1,111 @@
...
+    group = parser.getgroup("airflow")
+    group.addoption(
+        "--with-db-init",
+        action="store_true",
+        dest="db_init",
+        help="Forces database initialization before tests",
+    )
+
+
+@pytest.fixture(autouse=True, scope="session")
```
Comment (potiuk) [inline tests/conftest.py:63]: This is FANTASTIC!

**Review by potiuk — APPROVED** [review]
Fantastic changes!

**Review by feluelle — APPROVED** [review]
🚀

File: tests/conftest.py (hunk 28 above)
Comment (turbaszek) [reply tests/conftest.py:63]: Moreover if everything passes then there's no log from db.reset ✅

**Comment by potiuk** [comment]
One more thing - @nuclearpinguin - when we merge it, I think it makes sense to "warn" people and let them know they should expect changes in the way tests are run. 

Also It would be great to test how the new pytest works with IDE

BTW. It's super cool to be able to simply to just run 'pytest' and get the tests working out-of-the-box.

**Comment by turbaszek** [comment]
> 
> Also It would be great to test how the new pytest works with IDE
> 

I updated info in LOCAL_VIRTUALENV.rst and BREEZE.rst. I tested pytest with IDE and one change I added was to skip db.init when running locally to not mess with your local environment, WDYT @potiuk?

File: BREEZE.rst [hunk 29]
```diff
@@ -506,40 +506,47 @@ capturing the stdout of your process.
...
+   pytest --with-db-init tests/test_core.py::TestCore::test_check_operators -s --logging-level=DEBUG
 
-For a single test method, run:
+There are many available options for selecting specific test in pytest. Details could be found
+in official documentation but here are few basic examples:
 
 .. code-block:: bash
 
-   run-tests tests.core:TestCore.test_check_operators -- -s --logging-level=DEBUG
+    pytest -k "TestCore and not check"
```
Comment (kaxil) [inline BREEZE.rst]: I am not able to fully understand this !

File: setup.py [hunk 30]
```diff
@@ -284,23 +284,22 @@ def write_version(filename: str = os.path.join(*["airflow", "git_version"])):
     'dumb-init>=1.2.2',
     'flake8>=3.6.0',
     'flake8-colors',
+    'flaky==3.6.1',
```
Comment (kaxil) [inline setup.py]:
```suggestion
    'flaky~=3.6.1',
```

File: BREEZE.rst (hunk 29 above)
Comment (turbaszek) [reply BREEZE.rst]: @kaxil  as described below:
"The example above will run ``TestCore.test_bash_operator_kill``  but not ``TestCore.test_check_operators``."

If that's still unclear I can link there one more time to pytest documentation, WDYT?

Comment (kaxil) [reply BREEZE.rst]: So if I want to run all core but not bash operators would it be `pytest -k "TestCore and not bash"` ?

Comment (turbaszek) [reply BREEZE.rst]: Running `pytest -k "TestCore and not bash"` will run TestCore class but will skip tests of this class that includes bash in their names. For better performance (due to test collection) you should do:
`pytest tests/tests_core.py -k "TestCore and not bash"`. 

This is useful when used like this:
`pytest tests/tests_core.py -k "test_check_operators"`
to run single test.

Comment (kaxil) [reply BREEZE.rst]: Thanks, I understand it better now. Can we add that example in the instructions too :)

Comment (turbaszek) [reply BREEZE.rst]: Sure!

Comment (turbaszek) [reply BREEZE.rst]: Done

File: setup.py [hunk 31]
```diff
@@ -371,23 +371,22 @@ def write_version(filename: str = os.path.join(*["airflow", "git_version"])):
...
-    'nose-ignore-docstring==0.2',
-    'nose-timer',
     'parameterized',
     'paramiko',
     'pre-commit',
     'pylint~=2.3.1',  # to be upgraded after fixing https://github.com/PyCQA/pylint/issues/3123
                       # We should also disable checking docstring at the module level
     'pysftp',
+    'pytest',
+    'pytest-cov==2.8.1',
```
Comment (feluelle) [inline setup.py]: Why do you think it makes sense to pin it? (and not pytest?)

Personally I never had issues with pytest-cov.

File: tests/conftest.py (hunk 28 above)
Comment (feluelle) [reply tests/conftest.py:63]: This is great! 👍

File: tests/conftest.py [hunk 32]
```diff
@@ -0,0 +1,120 @@
...
+    else:
+        print(
+            "Skipping initializing of the DB as it was initialized already.\n"
+            "You can re-initialize the database by adding --with-db-init flag when running tests."
+        )
+
+    # Initialize kerberos
+    kerberos = os.environ.get("KRB5_KTNAME")
+    if kerberos:
+        subprocess.call(["kinit", "-kt", kerberos, "airflow"])
```
Comment (ashb) [inline tests/conftest.py:120]: This should probably be `check_call` - if the `kinit` fails this won't throw an error.

Comment (potiuk) [reply tests/conftest.py:120]: Good call. @nuclearpinguin -> something as a follow up (including the setting of pytest as default test runner.
